File: mobile-apps/alloy-ios/Alloy/Volume/VolumeViewModel.swift
import Foundation
import Observation

/// Canonical implementation of the Alloy UX principles. Every other
/// feature's view model is expected to follow this shape: optimistic
/// UI with decoupled intent/confirmed, throttled sends, trailing edge
/// on drag end. See `mobile-apps/specs/volume-control.md`.
@Observable
@MainActor
final class VolumeViewModel {

    // MARK: - Observable state

    var intent: Int = 0
    var confirmed: Int = 0
    var connectionState: ConnectionState = .disconnected
    var isDragging: Bool = false

    private(set) var range: VolumeRange

    // MARK: - Dependencies

    private let target: any VolumeTarget
    private var confirmedTask: Task<Void, Never>?
    private var connectionTask: Task<Void, Never>?

    // MARK: - Throttling

    private var pendingIntent: Int?
    private var throttleTask: Task<Void, Never>?
    /// Minimum spacing between outgoing `setVolume` calls while dragging.
    private let throttleInterval: Duration = .milliseconds(100)

    init(target: any VolumeTarget) {
        self.target = target
        self.range = target.volumeRange
    }

    // MARK: - Lifecycle

    func start() {
        // These Tasks inherit the MainActor context from the enclosing
        // @MainActor method, so all state mutations happen on the main
        // actor without any explicit `MainActor.run` hops.
        confirmedTask = Task { [weak self] in
            guard let self else { return }
            for await value in target.confirmedStream() {
                self.confirmed = value
                if !self.isDragging { self.intent = value }
                self.range = self.target.volumeRange
            }
        }
        connectionTask = Task { [weak self] in
            guard let self else { return }
            for await state in target.connectionStream() {
                self.connectionState = state
            }
        }
        Task { await target.connect() }
    }

    func stop() {
        confirmedTask?.cancel()
        connectionTask?.cancel()
        throttleTask?.cancel()
        Task { await target.disconnect() }
    }

    // MARK: - Events

    func onDragStart() {
        isDragging = true
    }

    func onDragChange(_ value: Int) {
        let clamped = range.clamp(value)
        intent = clamped
        scheduleThrottledSend(clamped)
    }

    func onDragEnd() {
        isDragging = false
        let final = intent
        throttleTask?.cancel()
        throttleTask = nil
        pendingIntent = nil
        Task { await target.setVolume(final) }
    }

    // MARK: - Throttling

    private func scheduleThrottledSend(_ value: Int) {
        pendingIntent = value
        if throttleTask != nil { return }
        throttleTask = Task { [weak self] in
            guard let self else { return }
            try? await Task.sleep(for: self.throttleInterval)
            // Still on MainActor thanks to @_inheritActorContext.
            let toSend = self.pendingIntent
            self.pendingIntent = nil
            self.throttleTask = nil
            if let toSend {
                await self.target.setVolume(toSend)
            }
        }
    }
}

File: mobile-apps/alloy-android/app/src/main/kotlin/com/alloyremote/alloy/volume/VolumeViewModel.kt
package com.alloyremote.alloy.volume

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.alloyremote.alloy.model.ConnectionState
import com.alloyremote.alloy.model.VolumeRange
import com.alloyremote.alloy.target.VolumeTarget
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * Canonical implementation of the Alloy UX principles. Every other
 * feature's view model is expected to follow this shape: optimistic
 * UI with decoupled intent/confirmed, throttled sends, trailing edge
 * on drag end. See `mobile-apps/specs/volume-control.md`.
 */
class VolumeViewModel(
    private val target: VolumeTarget
) : ViewModel() {

    private val _intent = MutableStateFlow(0)
    val intent: StateFlow<Int> = _intent.asStateFlow()

    private val _confirmed = MutableStateFlow(0)
    val confirmed: StateFlow<Int> = _confirmed.asStateFlow()

    private val _connectionState =
        MutableStateFlow<ConnectionState>(ConnectionState.Disconnected)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _isDragging = MutableStateFlow(false)
    val isDragging: StateFlow<Boolean> = _isDragging.asStateFlow()

    val range: VolumeRange get() = target.volumeRange

    private var pendingIntent: Int? = null
    private var throttleJob: Job? = null
    private val throttleIntervalMs: Long = 100

    init {
        viewModelScope.launch {
            target.confirmedFlow().collect { value ->
                _confirmed.value = value
                if (!_isDragging.value) _intent.value = value
            }
        }
        viewModelScope.launch {
            target.connectionFlow().collect { state ->
                _connectionState.value = state
            }
        }
        viewModelScope.launch { target.connect() }
    }

    override fun onCleared() {
        throttleJob?.cancel()
        viewModelScope.launch { target.disconnect() }
        super.onCleared()
    }

    fun onDragStart() {
        _isDragging.value = true
    }

    fun onDragChange(newValue: Int) {
        val clamped = range.clamp(newValue)
        _intent.value = clamped
        scheduleThrottledSend(clamped)
    }

    fun onDragEnd() {
        _isDragging.value = false
        val final = _intent.value
        throttleJob?.cancel()
        throttleJob = null
        pendingIntent = null
        viewModelScope.launch { target.setVolume(final) }
    }

    private fun scheduleThrottledSend(value: Int) {
        pendingIntent = value
        if (throttleJob != null) return
        throttleJob = viewModelScope.launch {
            delay(throttleIntervalMs)
            val toSend = pendingIntent
            pendingIntent = null
            throttleJob = null
            toSend?.let { target.setVolume(it) }
        }
    }
}
